Sources/Plasma/Apps/plClient/Mac-Cocoa/plMacDisplayHelper.mm

```diff
@@ -0,0 +1,147 @@
+/*==LICENSE==*
+
+CyanWorlds.com Engine - MMOG client, server and tools
+Copyright (C) 2011  Cyan Worlds, Inc.
+
+This program is free software: you can redistribute it and/or modify
+it under the terms of the GNU General Public License as published by
+the Free Software Foundation, either version 3 of the License, or
+(at your option) any later version.
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+Additional permissions under GNU GPL version 3 section 7
+
+If you modify this Program, or any covered work, by linking or
+combining it with any of RAD Game Tools Bink SDK, Autodesk 3ds Max SDK,
+NVIDIA PhysX SDK, Microsoft DirectX SDK, OpenSSL library, Independent
+JPEG Group JPEG library, Microsoft Windows Media SDK, or Apple QuickTime SDK
+(or a modified version of those libraries),
+containing parts covered by the terms of the Bink SDK EULA, 3ds Max EULA,
+PhysX SDK EULA, DirectX SDK EULA, OpenSSL and SSLeay licenses, IJG
+JPEG Library README, Windows Media SDK EULA, or QuickTime SDK EULA, the
+licensors of this Program grant you additional
+permission to convey the resulting work. Corresponding Source for a
+non-source form of such a combination shall include the source code for
+the parts of OpenSSL and IJG JPEG Library used as well as that of the covered
+work.
+
+You can contact Cyan Worlds, Inc. by email legal@cyan.com
+ or by snail mail at:
+      Cyan Worlds, Inc.
+      14617 N Newport Hwy
+      Mead, WA   99021
+
+*==LICENSE==*/
+
+#include "plPipeline.h"
+#include "plMacDisplayHelper.h"
+
+void plMacDisplayHelper::SetCurrentScreen(hsDisplayHndl display)
+{
+    NSScreen *nsScreen;
+    for (NSScreen *screen in [NSScreen screens]) {
+        NSDictionary *deviceDescription = [screen deviceDescription];
+        NSNumber *screenID = deviceDescription[@"NSScreenNumber"];
+        if ([screenID unsignedIntValue] == display) {
+            nsScreen = screen;
+        }
+    }
+    SetCurrentScreen(nsScreen);
+}
+
+void plMacDisplayHelper::SetCurrentScreen(NSScreen *screen)
```

Comment:
Would it make sense to refactor the looping over `[NSScreen screens]` into its own static function that takes in an `hsDisplayHndl` and returns an `NSScreen*`, and then have this method just take the `hsDisplayHndl`?

I'm wondering if there's a way to avoid needing to have `NSScreen` defined in the header file, because that precludes it from being included in any non-ObjC files.

Reply:
Let me think about it. I mentioned above to @Hoikas that this header should never be included by anyone but the client. plDisplayHelper is the clean version.

What these bits are circling around is the client needs to be able to tell the display helper when the game window changes displays, and the display helper needs to react to that. Thats why this interface exposes things like NSScreen (and the client has a window did change screen delegate message.)

I should probably decide what a likely path forward looks like for that, and then change the code to match.

Reply:
I'm going to try to put a pin in this one. @dpogue, @Hoikas - this feedback was given in multiple comments so I'm going to try to pull the discussion together here.

The line I'm going to draw is that plMacDisplayHelper is part of the Mac client (which is accurate to where this lives) and thus should never be touched by any other module. Display helpers should inherit from plDisplayHelper, but be owned by the client. It's a way for the client to inject display logic into the engine.

With that said - _my opinion is_ removing Obj-C from the header is unnecessary because this is part of the Obj-C client. And this header should not be imported in any other module.